src/components/OfflineSearch.ts

```diff
@@ -0,0 +1,173 @@
+        this.onClear = this.onClear.bind(this);
+        this.onSearchKeyDown = this.onSearchKeyDown.bind(this);
+    }
+
+    render() {
+        return createElement("div", {
+            className: `widget-offline-search ${this.props.showSearchBar}`
+        },
+            createElement(Alert, { message: this.state.alertMessage }),
+            createElement("div", { className: "search-container" },
+                createElement("span", { className: "glyphicon glyphicon-search" }),
+                createElement("input", {
+                    className: "form-control", placeholder: "Search",
+                    ref: input => this.searchInput = input as HTMLInputElement
+                }),
+                createElement("button", {
+                    className: `btn-transparent ${this.state.buttonVisibility}`,
+                    ref: button => this.searchButton = button as HTMLButtonElement
+                },
+                    createElement("span", { className: "glyphicon glyphicon-remove" })
+                )
+            )
+        );
+    }
+
+    componentDidMount() {
+        this.findTargetNode();
+        this.isValidWidget();
+        this.setUpEvents();
+        this.setDefaultSearch();
+    }
+
+    componentWillUnmount() {
+        this.searchButton.removeEventListener("click", this.onClear);
+        this.searchInput.removeEventListener("keyup", this.onSearchKeyDown);
+    }
+
+    private findTargetNode() {
+        let queryNode = findDOMNode(this).parentNode as HTMLElement;
+        let targetNode: HTMLElement | null = null;
+        const targetName = this.props.targetGridName;
+
+        while (!targetNode) {
+            targetNode = queryNode.querySelector(`.mx-name-${targetName}`) as HTMLElement;
+            if (window.document.isEqualNode(queryNode)) break;
+            queryNode = queryNode.parentNode as HTMLElement;
+        }
+
+        if (!targetNode) {
+            this.setState({ alertMessage: `search offline widget: unable to find grid with the name "${targetName}"` });
+        } else {
+            this.targetWidget = dijitRegistry.byNode(targetNode);
+        }
+    }
+
+    private isValidWidget(): boolean {
+        if (this.targetWidget && this.targetWidget.declaredClass === "mxui.widget.ListView") {
+            if (this.targetWidget._onLoad
+                && this.targetWidget._loadMore
+                && this.targetWidget._renderData
+                && this.targetWidget._datasource
+                && this.targetWidget._datasource.atEnd
+                && typeof this.targetWidget._datasource._pageSize !== "undefined"
+                && (typeof this.targetWidget._datasource._setsize !== "undefined"
+                || typeof this.targetWidget._datasource._setSize !== "undefined")) {
+                    return true;
+            } else {
+                this.setState({ alertMessage: "search offline widget: this Mendix version is incompatible with the offline search widget" });
```

Comment:
returning message will change the entire working of function. because it is used in validations